File: typescript/src/types/binary-serializer-string.ts
import { BinaryTypeSerializer } from './binary-type-serializer.interface';
import { DataReader } from '../buffers/data-reader.interface';
import { DataWriter } from '../buffers/data-writer.interface';
import { BinarySerializerUShort } from './binary-serializer-ushort';
import {InvalidStringLengthException} from "./exceptions/invalid-string-length-exception";

/**
 * Reads (and writes) string values to (and from) a binary data source.
 *
 * @public
 * @exported
 * @implements {BinaryTypeSerializer<string | null>}
 */
export class BinarySerializerString implements BinaryTypeSerializer<string | null> {
    private static readonly MAX_STRING_LENGTH_IN_BYTES = 2**16 - 1;
    private readonly binarySerializerUShort: BinarySerializerUShort;
    private readonly encoding: typeof TextEncoder;

    constructor(encoding: typeof TextEncoder = TextEncoder) {
        this.binarySerializerUShort = new BinarySerializerUShort();
        this.encoding = encoding;
    }

    /**
     * @throws {InvalidStringLengthException} The string length exceeds the maximum allowed length.
     */
    encode(writer: DataWriter, value: string | null): void {
        if (value === null) {
            this.writeNullFlag(writer, true);
            return;
        }

        this.writeNullFlag(writer, false);

        const encoder = new this.encoding();
        const bytes = encoder.encode(value);

        if (bytes.length > BinarySerializerString.MAX_STRING_LENGTH_IN_BYTES) {
            throw new InvalidStringLengthException(bytes.length, BinarySerializerString.MAX_STRING_LENGTH_IN_BYTES);
        }

        this.binarySerializerUShort.encode(writer, bytes.length);
        writer.writeBytes(bytes);
    }
  
    decode(reader: DataReader): string | null {
        if (this.readNullFlag(reader)) {
            return null;
        }

        const length = this.binarySerializerUShort.decode(reader);
        const bytes = reader.readBytes(length);

        return new TextDecoder().decode(bytes);
    }

    getEquals(a: string | null, b: string | null): boolean {
        if (a === null && b === null) {
            return true;
        }

        if (a === null || b === null) {
            return false;
        }

        return a === b;
    }

    private readNullFlag(reader: DataReader): boolean {
        return reader.readBit();
    }

    private writeNullFlag(writer: DataWriter, flag: boolean): void {
        writer.writeBit(flag);
    }
}

File: dotnet/Barchart.BinarySerializer/Types/BinarySerializerString.cs
#region Using Statements

using System.Text;

using Barchart.BinarySerializer.Buffers;
using Barchart.BinarySerializer.Types.Exceptions;
using Barchart.BinarySerializer.Utilities;

#endregion

namespace Barchart.BinarySerializer.Types;

/// <summary>
///     Reads (and writes) string values to (and from) a binary data source.
/// </summary>
public class BinarySerializerString : IBinaryTypeSerializer<string?>
{
    #region Constants

    private const int MAXIMUM_STRING_LENGTH_IN_BYTES = ushort.MaxValue;

    #endregion

    #region Fields
    
    private static readonly Encoding DEFAULT_ENCODING = Encoding.UTF8;
    private readonly Encoding _encoding;

    private readonly BinarySerializerUShort _binarySerializerUShort;

    #endregion

    #region Constructor(s)

    /// <summary>
    ///     Initializes a new instance of the <see cref="BinarySerializerString"/> class using the default encoding.
    /// </summary>
    public BinarySerializerString() : this(DEFAULT_ENCODING)
    {

    }

    /// <summary>
    ///     Initializes a new instance of the <see cref="BinarySerializerString"/> class using the specified encoding.
    /// </summary>
    /// <param name="encoding">
    ///     The encoding to use when reading and writing strings.
    /// </param>
    public BinarySerializerString(Encoding encoding)
    {
        _encoding = encoding;

        _binarySerializerUShort = new BinarySerializerUShort();
    }

    #endregion

    #region Methods

    /// <inheritdoc />
    /// <exception cref="InvalidStringLengthException">
    ///     Thrown when the serialized string would require more than <see cref="MAXIMUM_STRING_LENGTH_IN_BYTES"/> bytes.
    /// </exception>
    public void Encode(IDataBufferWriter writer, string? value)
    {
        Serialization.WriteNullFlag(writer, value == null);
        
        if (value == null)
        {
            return;
        }

        byte[] bytes = _encoding.GetBytes(value);

        if (bytes.Length > MAXIMUM_STRING_LENGTH_IN_BYTES)
        {
            throw new InvalidStringLengthException(bytes.Length, MAXIMUM_STRING_LENGTH_IN_BYTES);
        }

        _binarySerializerUShort.Encode(writer, Convert.ToUInt16(bytes.Length));

        writer.WriteBytes(bytes);
    }

    /// <inheritdoc />
    public string? Decode(IDataBufferReader reader)
    {
        if (Serialization.ReadNullFlag(reader))
        {
            return null;
        }

        return _encoding.GetString(reader.ReadBytes(_binarySerializerUShort.Decode(reader)));
    }

    /// <inheritdoc />
    public bool GetEquals(string? a, string? b)
    {
        if (a == null && b == null)
        {
            return true;
        }

        if (a == null || b == null)
        {
            return false;
        }

        return a.Equals(b);
    }

    #endregion
}